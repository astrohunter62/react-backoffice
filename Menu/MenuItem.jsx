import React from 'react'
import PropTypes from 'prop-types'

import { ListItem, ListItemIcon, ListItemText } from 'material-ui/List'

const MenuItem = ({ redirectTo, url, title, icon: Icon }) => {
  return (
    <ListItem button onClick={() => redirectTo(url)}>
      {Icon? (
        <ListItemIcon>
          <Icon />
        </ListItemIcon>
      ) : null}
      <ListItemText primary={title} />
    </ListItem>
  )
}

MenuItem.propTypes = {
  redirectTo: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  icon: PropTypes.func,
}

export default MenuItem
