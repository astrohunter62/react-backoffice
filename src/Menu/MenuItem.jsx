import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import withStyles from 'material-ui/styles/withStyles'
import { ListItem, ListItemIcon, ListItemText } from 'material-ui/List'

const styles = {
  disabled: {
    opacity: 0.5,
    pointerEvents: 'none',
  },
}

const MenuItem = ({
  redirectTo,
  url,
  title,
  icon: Icon,
  disabled,
  classes,
}) => (
  <ListItem
    button
    onClick={() => redirectTo(url)}
    className={classNames({
        [classes.disabled]: disabled,
      })}
  >
    {Icon ? (
      <ListItemIcon>
        <Icon />
      </ListItemIcon>
      ) : null}
    <ListItemText primary={title} />
  </ListItem>
)

MenuItem.propTypes = {
  redirectTo: PropTypes.func.isRequired,
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  icon: PropTypes.func,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
}

MenuItem.defaultProps = {
  disabled: false,
  icon: null,
}

export default withStyles(styles)(MenuItem)
