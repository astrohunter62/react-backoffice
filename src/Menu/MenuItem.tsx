import React, { FunctionComponent } from "react";
import classNames from "classnames";
import {
  ListItem,
  ListItemIcon,
  ListItemText,
  makeStyles,
} from "@material-ui/core";

const useStyles = makeStyles(() => ({
  isDisabled: {
    opacity: 0.5,
    pointerEvents: "none",
  },
}));

type MenuItemProps = {
  redirectTo: (...args: any[]) => any;
  url: string;
  title: string;
  isDisabled?: boolean;
  icon?: JSX.Element;
  className?: string;
};

const MenuItem: FunctionComponent<MenuItemProps> = ({
  redirectTo,
  url,
  title,
  icon,
  isDisabled = false,
  className,
}) => {
  const classes = useStyles();

  return (
    <ListItem
      button
      onClick={() => redirectTo(url)}
      className={classNames(className, {
        [classes.isDisabled]: isDisabled,
      })}
    >
      {icon && <ListItemIcon>{icon}</ListItemIcon>}

      <ListItemText primary={title} />
    </ListItem>
  );
};

export default MenuItem;
