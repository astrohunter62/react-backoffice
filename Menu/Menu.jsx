import React from 'react'
import PropTypes from 'prop-types'
import { withStyles } from 'material-ui/styles'
import List from 'material-ui/List'
import Divider from 'material-ui/Divider'

import MenuItem from './MenuItem'

const styles = theme => ({
  root: {
    width: '100%',
    maxWidth: 360,
    background: theme.palette.background.paper,
  },
})

const Menu = ({ data, redirectTo, classes }) => {
  return (
    <div className={classes.root}>
      <List>
        {data.map((item, index) => {
          switch (item.type) {
            case 'divider':
              return <Divider key={`menu-${index}`} />
              break

            default:
              return (
                <MenuItem
                  key={`menu-${index}`}
                  redirectTo={redirectTo}
                  url={item.url}
                  title={item.title}
                  icon={item.icon}
                />
              )
              break
          }
        })}
      </List>
    </div>
  )
}

Menu.propTypes = {
  redirectTo: PropTypes.func.isRequired,
  classes: PropTypes.object.isRequired,
}

export default withStyles(styles)(Menu)
