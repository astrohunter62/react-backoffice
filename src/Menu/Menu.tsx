import React, { Fragment, FunctionComponent } from "react";
import { Divider, List, makeStyles, Theme } from "@material-ui/core";
import classNames from "classnames";
import MenuItem from "./MenuItem";

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    width: "100%",
    maxWidth: 360,
    background: theme.palette.background.paper,
  },

  submenu: {
    paddingTop: 0,
  },
}));

type MenuItemLink = {
  type: "link";
  url: string;
  title: string;
  isDisabled?: boolean;
  icon?: JSX.Element;
  items?: MenuItemLink[];
};

type MenuItemDivider = {
  type: "divider";
};

export type MenuDataItem = MenuItemLink | MenuItemDivider;

type MenuProps = {
  redirectTo: (...args: any[]) => any;
  data: MenuDataItem[];
  className?: string;
  isDense?: boolean;
};

const Menu: FunctionComponent<MenuProps> = ({
  data,
  redirectTo,
  isDense = false,
  className,
}) => {
  const classes = useStyles();

  return (
    <div className={classNames(classes.root, className)}>
      <List dense={isDense} className={isDense ? classes.submenu : ""}>
        {data.map((item, index) => {
          switch (item.type) {
            case "divider":
              return <Divider key={`menu-${index}`} />;
            default:
              return (
                <Fragment key={`menu-${index}`}>
                  <MenuItem
                    redirectTo={redirectTo}
                    url={item.url}
                    title={item.title}
                    isDisabled={item.isDisabled}
                    icon={item.icon}
                  />

                  {item.items && (
                    <Menu data={item.items} redirectTo={redirectTo} isDense />
                  )}
                </Fragment>
              );
          }
        })}
      </List>
    </div>
  );
};

export default Menu;
