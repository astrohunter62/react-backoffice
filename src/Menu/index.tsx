import React, { Fragment, FunctionComponent } from "react";
import { Divider, List, makeStyles, Theme } from "@material-ui/core";
import classnames from "classnames";
import MenuItem from "./MenuItem";

const useStyles = makeStyles((theme: Theme) => ({
  root: {
    width: "100%",
    maxWidth: 360,
    background: theme.palette.background.paper,
  },

  submenu: {
    paddingTop: 0,
  },

  subitem: {
    paddingLeft: theme.spacing(4),
  },

  divider: {
    marginTop: theme.spacing(1),
    marginBottom: theme.spacing(1),
  },
}));

type MenuItemBase = {
  title: string;
  isDisabled?: boolean;
  icon?: JSX.Element;
  items?: MenuItemLink[];
};
type MenuItemLabel = MenuItemBase & {
  type: "label";
};

type MenuItemLink = MenuItemBase & {
  type: "link";
  url: string;
};

type MenuItemDivider = {
  type: "divider";
};

export type MenuDataItem = MenuItemLink | MenuItemLabel | MenuItemDivider;

type MenuProps = {
  redirectTo: (...args: any[]) => any;
  data: MenuDataItem[];
  className?: string;
  isDense?: boolean;
  menuItemClassName?: string;
};

const Menu: FunctionComponent<MenuProps> = ({
  data,
  redirectTo,
  isDense = false,
  menuItemClassName,
  className,
}) => {
  const classes = useStyles();

  return (
    <div className={classnames(classes.root, className)}>
      <List
        dense={isDense}
        className={classnames({ [classes.submenu]: isDense })}
      >
        {data.map((item, index) => {
          switch (item.type) {
            case "divider":
              return (
                <Divider key={`menu-${index}`} className={classes.divider} />
              );
            default:
              return (
                <Fragment key={`menu-${index}`}>
                  <MenuItem
                    redirectTo={redirectTo}
                    url={(item as any).url}
                    title={item.title}
                    isButton={item.type === "link"}
                    isDisabled={item.isDisabled}
                    icon={item.icon}
                    className={menuItemClassName}
                  />

                  {item.items && (
                    <Menu
                      data={item.items}
                      redirectTo={redirectTo}
                      isDense
                      menuItemClassName={classes.subitem}
                    />
                  )}
                </Fragment>
              );
          }
        })}
      </List>
    </div>
  );
};

export default Menu;
