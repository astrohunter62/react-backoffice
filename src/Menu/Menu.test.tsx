import React from "react";
import { shallow } from "enzyme";

import Icon from "@material-ui/icons/AccessAlarm";

import Menu from ".";

it("renders correctly", () => {
  const tree = shallow(
    <Menu
      data={[
        {
          type: "link",
          title: "Title",
          url: "/",
          icon: <Icon />,
        },
      ]}
      redirectTo={() => {}}
    />,
  );

  expect(tree).toMatchSnapshot();
});
