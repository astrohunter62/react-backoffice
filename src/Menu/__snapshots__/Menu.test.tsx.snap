// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<WithStyles(Menu)
  data={
    Array [
      Object {
        "icon": <UNDEFINED />,
        "title": "Title",
        "type": "link",
        "url": "/",
      },
    ]
  }
  redirectTo={[Function]}
>
  <Menu
    classes={
      Object {
        "root": "Menu-root-1",
      }
    }
    data={
      Array [
        Object {
          "icon": <UNDEFINED />,
          "title": "Title",
          "type": "link",
          "url": "/",
        },
      ]
    }
    redirectTo={[Function]}
  >
    <div
      className="Menu-root-1"
    >
      <WithStyles(ForwardRef(List))>
        <ForwardRef(List)
          classes={
            Object {
              "dense": "MuiList-dense",
              "padding": "MuiList-padding",
              "root": "MuiList-root",
              "subheader": "MuiList-subheader",
            }
          }
        >
          <ul
            className="MuiList-root MuiList-padding"
          >
            <WithStyles(MenuItem)
              icon={<UNDEFINED />}
              key="menu-0"
              redirectTo={[Function]}
              title="Title"
              url="/"
            >
              <MenuItem
                classes={
                  Object {
                    "isDisabled": "MenuItem-isDisabled-6",
                  }
                }
                icon={<UNDEFINED />}
                isDisabled={false}
                redirectTo={[Function]}
                title="Title"
                url="/"
              >
                <WithStyles(ForwardRef(ListItem))
                  button={true}
                  className=""
                  onClick={[Function]}
                >
                  <ForwardRef(ListItem)
                    button={true}
                    className=""
                    classes={
                      Object {
                        "alignItemsFlexStart": "MuiListItem-alignItemsFlexStart",
                        "button": "MuiListItem-button",
                        "container": "MuiListItem-container",
                        "dense": "MuiListItem-dense",
                        "disabled": "Mui-disabled",
                        "divider": "MuiListItem-divider",
                        "focusVisible": "Mui-focusVisible",
                        "gutters": "MuiListItem-gutters",
                        "root": "MuiListItem-root",
                        "secondaryAction": "MuiListItem-secondaryAction",
                        "selected": "Mui-selected",
                      }
                    }
                    onClick={[Function]}
                  >
                    <WithStyles(ForwardRef(ButtonBase))
                      className="MuiListItem-root MuiListItem-gutters MuiListItem-button"
                      component="div"
                      disabled={false}
                      focusVisibleClassName="Mui-focusVisible"
                      onClick={[Function]}
                    >
                      <ForwardRef(ButtonBase)
                        className="MuiListItem-root MuiListItem-gutters MuiListItem-button"
                        classes={
                          Object {
                            "disabled": "Mui-disabled",
                            "focusVisible": "Mui-focusVisible",
                            "root": "MuiButtonBase-root",
                          }
                        }
                        component="div"
                        disabled={false}
                        focusVisibleClassName="Mui-focusVisible"
                        onClick={[Function]}
                      >
                        <div
                          aria-disabled={false}
                          className="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
                          onBlur={[Function]}
                          onClick={[Function]}
                          onDragLeave={[Function]}
                          onFocus={[Function]}
                          onKeyDown={[Function]}
                          onKeyUp={[Function]}
                          onMouseDown={[Function]}
                          onMouseLeave={[Function]}
                          onMouseUp={[Function]}
                          onTouchEnd={[Function]}
                          onTouchMove={[Function]}
                          onTouchStart={[Function]}
                          role="button"
                          tabIndex={0}
                        >
                          <WithStyles(ForwardRef(ListItemIcon))
                            key=".0"
                          >
                            <ForwardRef(ListItemIcon)
                              classes={
                                Object {
                                  "alignItemsFlexStart": "MuiListItemIcon-alignItemsFlexStart",
                                  "root": "MuiListItemIcon-root",
                                }
                              }
                            >
                              <div
                                className="MuiListItemIcon-root"
                              >
                                <ForwardRef>
                                  <WithStyles(ForwardRef(SvgIcon))>
                                    <ForwardRef(SvgIcon)
                                      classes={
                                        Object {
                                          "colorAction": "MuiSvgIcon-colorAction",
                                          "colorDisabled": "MuiSvgIcon-colorDisabled",
                                          "colorError": "MuiSvgIcon-colorError",
                                          "colorPrimary": "MuiSvgIcon-colorPrimary",
                                          "colorSecondary": "MuiSvgIcon-colorSecondary",
                                          "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                                          "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                                          "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                                          "root": "MuiSvgIcon-root",
                                        }
                                      }
                                    >
                                      <svg
                                        aria-hidden={true}
                                        className="MuiSvgIcon-root"
                                        focusable="false"
                                        viewBox="0 0 24 24"
                                      >
                                        <path
                                          d="M22 5.72l-4.6-3.86-1.29 1.53 4.6 3.86L22 5.72zM7.88 3.39L6.6 1.86 2 5.71l1.29 1.53 4.59-3.85zM12.5 8H11v6l4.75 2.85.75-1.23-4-2.37V8zM12 4c-4.97 0-9 4.03-9 9s4.02 9 9 9c4.97 0 9-4.03 9-9s-4.03-9-9-9zm0 16c-3.87 0-7-3.13-7-7s3.13-7 7-7 7 3.13 7 7-3.13 7-7 7z"
                                        />
                                      </svg>
                                    </ForwardRef(SvgIcon)>
                                  </WithStyles(ForwardRef(SvgIcon))>
                                </ForwardRef>
                              </div>
                            </ForwardRef(ListItemIcon)>
                          </WithStyles(ForwardRef(ListItemIcon))>
                          <WithStyles(ForwardRef(ListItemText))
                            key=".1"
                            primary="Title"
                          >
                            <ForwardRef(ListItemText)
                              classes={
                                Object {
                                  "dense": "MuiListItemText-dense",
                                  "inset": "MuiListItemText-inset",
                                  "multiline": "MuiListItemText-multiline",
                                  "primary": "MuiListItemText-primary",
                                  "root": "MuiListItemText-root",
                                  "secondary": "MuiListItemText-secondary",
                                }
                              }
                              primary="Title"
                            >
                              <div
                                className="MuiListItemText-root"
                              >
                                <WithStyles(ForwardRef(Typography))
                                  className="MuiListItemText-primary"
                                  component="span"
                                  display="block"
                                  variant="body1"
                                >
                                  <ForwardRef(Typography)
                                    className="MuiListItemText-primary"
                                    classes={
                                      Object {
                                        "alignCenter": "MuiTypography-alignCenter",
                                        "alignJustify": "MuiTypography-alignJustify",
                                        "alignLeft": "MuiTypography-alignLeft",
                                        "alignRight": "MuiTypography-alignRight",
                                        "body1": "MuiTypography-body1",
                                        "body2": "MuiTypography-body2",
                                        "button": "MuiTypography-button",
                                        "caption": "MuiTypography-caption",
                                        "colorError": "MuiTypography-colorError",
                                        "colorInherit": "MuiTypography-colorInherit",
                                        "colorPrimary": "MuiTypography-colorPrimary",
                                        "colorSecondary": "MuiTypography-colorSecondary",
                                        "colorTextPrimary": "MuiTypography-colorTextPrimary",
                                        "colorTextSecondary": "MuiTypography-colorTextSecondary",
                                        "displayBlock": "MuiTypography-displayBlock",
                                        "displayInline": "MuiTypography-displayInline",
                                        "gutterBottom": "MuiTypography-gutterBottom",
                                        "h1": "MuiTypography-h1",
                                        "h2": "MuiTypography-h2",
                                        "h3": "MuiTypography-h3",
                                        "h4": "MuiTypography-h4",
                                        "h5": "MuiTypography-h5",
                                        "h6": "MuiTypography-h6",
                                        "noWrap": "MuiTypography-noWrap",
                                        "overline": "MuiTypography-overline",
                                        "paragraph": "MuiTypography-paragraph",
                                        "root": "MuiTypography-root",
                                        "srOnly": "MuiTypography-srOnly",
                                        "subtitle1": "MuiTypography-subtitle1",
                                        "subtitle2": "MuiTypography-subtitle2",
                                      }
                                    }
                                    component="span"
                                    display="block"
                                    variant="body1"
                                  >
                                    <span
                                      className="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
                                    >
                                      Title
                                    </span>
                                  </ForwardRef(Typography)>
                                </WithStyles(ForwardRef(Typography))>
                              </div>
                            </ForwardRef(ListItemText)>
                          </WithStyles(ForwardRef(ListItemText))>
                          <WithStyles(memo)
                            center={false}
                          >
                            <ForwardRef(TouchRipple)
                              center={false}
                              classes={
                                Object {
                                  "child": "MuiTouchRipple-child",
                                  "childLeaving": "MuiTouchRipple-childLeaving",
                                  "childPulsate": "MuiTouchRipple-childPulsate",
                                  "ripple": "MuiTouchRipple-ripple",
                                  "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                                  "rippleVisible": "MuiTouchRipple-rippleVisible",
                                  "root": "MuiTouchRipple-root",
                                }
                              }
                            >
                              <span
                                className="MuiTouchRipple-root"
                              >
                                <TransitionGroup
                                  childFactory={[Function]}
                                  component={null}
                                  exit={true}
                                />
                              </span>
                            </ForwardRef(TouchRipple)>
                          </WithStyles(memo)>
                        </div>
                      </ForwardRef(ButtonBase)>
                    </WithStyles(ForwardRef(ButtonBase))>
                  </ForwardRef(ListItem)>
                </WithStyles(ForwardRef(ListItem))>
              </MenuItem>
            </WithStyles(MenuItem)>
          </ul>
        </ForwardRef(List)>
      </WithStyles(ForwardRef(List))>
    </div>
  </Menu>
</WithStyles(Menu)>
`;
